### PR TITLE
New section for DocBlock + a couple of new standards

### PR DESCRIPTION
**Pull request by mathieupinet**
New standards: 
* single-line & multi-line class variable blocks
* array of things

### CODE REVIEW

**Comment by sjparkinson** [comment]
Needs some links to what tags are actually valid ([how about this?](http://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_phpDocumentor.howto.pkg.html#basics.tags)), and what are **required** on a method.

I'd assume `param`, `returns`, `throws`, what else?

What about `{@inheritDoc}` usage?

I'd say we _should_ recommend having a distinction between a short and long description on a method, kinda like git.

What about the formatting of the (long) description? I'm cool with markdown, but maybe tools better support html.

File: standards.md [hunk 1]
```diff
@@ -57,36 +43,72 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
...
+      /**
+       * The Bar used to fight the foo
+       * @var Bar
+       */
+      private $bar;
+  }
+  ```
+
+5. DocBlock types for scalar values MUST be one of: `bool` (not boolean), `int` (not integer), `string`, `float` (not double).
+6. DocBlock types for parameters/return values that are arrays of a single type MUST be written as: `Type[]`. If the variable is a multi-dimensional array, it MUST be represented with one set of `[]` per depth (`int[][]` for an array of array of ints)
```
Comment (brendankay) [inline standards.md]: Looking at this format, my instinct is that the arrays are numeric.
Not sure how best to represent associative arrays.

**Comment by mathieupinet** [comment]
@sjparkinson `@return` and `@param` are the only required ones, isn't it clear from https://github.com/graze/standards/pull/21/files#diff-04c6e90faac2675aa89e2176d2eec7d8R66 ?

File: standards.md (hunk 1 above)
Comment (mathieupinet) [inline standards.md]: as in:
```
$foo['apple']['blue'] = new Foo(1);
$foo['apple']['yellow'] = new Foo(2);
$foo['pear']['yellow'] = new Foo(3);
```
?

Should I just specify that the rule applies to both associative and sequential arrays?

Comment (brendankay) [inline standards.md]: I'm not sure anything needs to change just wondering how useful defining something as multidimensional is, I guess it shows that you need to dig deeper...

**Comment by mathieupinet** [comment]
@sjparkinson thinking about `@inheritdoc` (and learning that it only really inherits the description) led me to this thought: should classes that inherit the exact same documentation even HAVE a DocBlock?

**Comment by wpillar** [comment]
I've been thinking about `@inheritdoc` and I'm currently thinking that we shouldn't use it and we should explicitly define the doc block for sub-methods.

**Comment by mathieupinet** [comment]
@sjparkinson I haven't come across methods with a description verbose enough that I had to consider a special format for it. Is it an issue in our Open Source projects?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,261 @@
...
+
+**example of a method's DocBlock:**
+```
+    /**
+     * Foos the bars
+     *
+     * @param int $barId Some number
+     * @param Bar[] $allTheBars Collection of Bar objects
+     * @return string[] List of messages
+     *
```
Comment (sjparkinson) [inline README.md]: Why the empty line? :fire:

File: README.md [hunk 3]
```diff
@@ -1,3 +1,261 @@
...
+
+5. DocBlock types for scalar values MUST be one of: `bool` (not boolean), `int` (not integer), `string`, `float` (not double).
+6. DocBlock types for parameters/return values that are arrays of a single type MUST be written as: `Type[]`. If the variable is a multi-dimensional array, it MUST be represented with one set of `[]` per depth (`int[][]` for an array of array of ints)
+7. If the parameter/return value is an array of several types, it MUST be described in the DocBlock as `mixed[]`.
+8. If the parameter/return value is an array that is sometimes empty, it MUST still be written as in 6) (NOT `Type[]|[]`)
+
+**example of a method's DocBlock:**
+```
+    /**
+     * Foos the bars
```
Comment (sjparkinson) [inline README.md]: While perhaps this shouldn't be enforced, it'd be nice if the example short description ended with a full-stop.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,261 @@
...
+     *
+     * @param int $barId Some number
+     * @param Bar[] $allTheBars Collection of Bar objects
+     * @return string[] List of messages
```
Comment (wpillar) [inline README.md]: How are we feeling about a blank line between `@param` tags and `@return`?

File: README.md [hunk 5]
```diff
@@ -1,3 +1,261 @@
...
+      
+      /**
+       * The Bar used to fight the foo
+       * @var Bar
+       */
+      private $bar;
+  }
+  ```
+
+5. DocBlock types for scalar values MUST be one of: `bool` (not boolean), `int` (not integer), `string`, `float` (not double).
```
Comment (sjparkinson) [inline README.md]: This kinda thing we should enforce in the style check or it'll never happen.

Comment (wpillar) [inline README.md]: It already is for `bool` and `int` :speak_no_evil:

**Comment by sjparkinson** [comment]
> @sjparkinson @return and @param are the only required ones, isn't it clear from https://github.com/graze/standards/pull/21/files#diff-04c6e90faac2675aa89e2176d2eec7d8R66 ?

Yes, after re-reading it. Being totally honest I found a lot of these verbose and just skimmed them.

> @sjparkinson thinking about @inheritdoc (and learning that it only really inherits the description) led me to this thought: should classes that inherit the exact same documentation even HAVE a DocBlock?

Yes, return values can change, and I've never seen otherwise elsewhere.

> @sjparkinson I haven't come across methods with a description verbose enough that I had to consider a special format for it. Is it an issue in our Open Source projects?

Something to consider another time.

File: README.md (hunk 4 above)
Comment (mathieupinet) [inline README.md]: I usually go for no blank line, but don't think it impedes/helps readability.

File: README.md (hunk 2 above)
Comment (mathieupinet) [inline README.md]: I usually separate out the description and parameters/returns. I don't think it should be enforced either way, do you?

**Comment by mathieupinet** [comment]
Closed to create a new PR as #25.

**Pull request closed without merging**